query-planner-js/src/buildPlan.ts

```diff
@@ -3351,23 +3377,19 @@ function handleRequires(
     }
 
     // We know the @require needs createdGroups. We do want to know however if any of the conditions was
-    // fetched from our `newGroup`. If not, then this means that `createdGroup` don't really depend on
-    // the current `group`, but can be dependencies of the parent (or even merged into this parent).
-    // To know this, we check if `newGroup` inputs contains its inputs (meaning the fetch is
-    // useless: we jump to it but didn't get anything new). Not that this isn't perfect because
-    // in the case of multiple keys between `newGroup` and its parent, we could theoretically take a
-    // different key on the way in that on the way back. In other words, `newGroup` selection may only
-    // be fetching a key that happens to not be the one in its inputs, and in that case the code below
-    // will not remove `newGroup` even though it would be more efficient to do so. Handling this properly
-    // is more complex however and it's sufficiently unlikely to happpen that we ignore that "optimization"
-    // for now. If someone run into this and notice, we can optimize then.
+    // fetched from our `newGroup`. If not, then this means that the `createdGroups` don't really depend on
+    // the current `group` and can be dependencies of the parent (or even merged into this parent).
+    //
+    // So we want to know if anything in `newGroup` selection cannot be fetched directly from the parent.
+    // For that, we first remove any of `newGroup` inputs from its selection: in most case, `newGroup`
+    // will just contain the key needed to jump back to its parent, and those would usually be the same
+    // as the inputs. And since by definition we know `newGroup`'s inputs are already fetched, we
+    // know they are not things that we need. Then, we check if what remains (often empty) can be
+    // directly fetched from the parent. If it can, then we can just merge `newGroup` into that parent.
+    // Otherwise, we will have to "keep it".
     // Note: it is to be sure this test is not poluted by other things in `group` that we created `newGroup`.
-    // Note2: `__typename` selections adds a bit of complexity. That is, if `newGroup` selection is not
-    // strictly contained in its inputs but only due to the selection of some `__typename`, then we
-    // still want to ignore that group, because `__typename` are always trivially queriable from any
-    // type in any subgraph and so that `__typename` can always be fetched from the parent. Which is
-    // what the `newGroupIsUnneeded` check ignores `__typename` in the selection.
-    const newGroupIsUnneeded = newGroup.inputs!.contains(withoutTypename(newGroup.selection)) && parent.path;
+    newGroup.removeInputsFromSelection();
```

Comment:
Note that this is the core change here. The rest of the diff is mostly about supporting this change.

And the problem was that we could have cases where `newGroupIsUnneeded` was true, but some of selections of `newGroup` are not valid selections for its parent (despite the selections of `newGroup` being essentially its inputs, there is cases where the direct parent is not the group that has fetched those inputs). So when we tried to call `parent.group.mergeChildIn(newGroup)` below, this was actually throwing an assertion, and for good reason.

The change here make it so that we only try to merge `newGroup` to its parent if we can guarantee that the merge will not throw.